core/Tracker/Visit/VisitProperties.php

```diff
@@ -52,27 +63,72 @@ public function &getProperties()
      *
      * @param string $name The property name.
      * @param mixed $value The property value.
+     *
+     * @return void
      */
-    public function setProperty($name, $value)
+    public function setProperty($name, $value): void
     {
         $this->visitInfo[$name] = $value;
     }
 
     /**
      * Unsets all visit properties.
+     *
+     * @return void
      */
-    public function clearProperties()
+    public function clearProperties(): void
     {
-        $this->visitInfo = array();
+        $this->visitInfo = [];
     }
 
     /**
      * Sets all visit properties.
      *
      * @param array $properties
+     *
+     * @return void
      */
-    public function setProperties($properties)
+    public function setProperties(array $properties): void
     {
         $this->visitInfo = $properties;
     }
+
+    /**
+     * Set the initial value of a property, this should remain unchanged throughout request processing.
+     *
+     * @param string $name
+     * @param mixed $value
+     *
+     * @return void
+     */
+    public function initializeImmutableProperty(string $name, $value): void
+    {
+        if (isset($this->visitInfoImmutableProperties[$name])) {
+            throw new \Exception('Immutable visit properties cannot be modified');
+        }
+        $this->visitInfoImmutableProperties[$name] = $value;
+    }
+
+    /**
+     * Returns a visit property, unmodified by request processors. Returns `null` if not set.
+     *
+     * @param string $name The property name.
+     *
+     * @return mixed|null
+     */
+    public function getImmutableProperty(string $name)
+    {
+        return isset($this->visitInfoImmutableProperties[$name]) ? $this->visitInfoImmutableProperties[$name] : null;
```

Comment:
Slightly shorter equivalent 
```suggestion
        return $this->visitInfoImmutableProperties[$name]) ?? null;
```